src/base.mo

```diff
@@ -0,0 +1,424 @@
+
+    assert Nat64.bitcountNonZero(root_aridity_) == 1; // 2-power
+    public let root_bitlength_ = Nat64.bitcountTrailingZero(root_aridity_);
+    assert root_bitlength_ > 0 and root_bitlength_ % bitlength_ == 0; // => root_bitlength_ >= bitlength_
+    assert root_bitlength_ <= key_size_ * 8;
+
+    let root_depth = Nat32.toNat16(Nat64.toNat32(root_bitlength_ / bitlength_));
+    public let root_bitlength = Nat32.toNat16(Nat64.toNat32(root_bitlength_));
+
+    public let node_size : Nat64 = aridity_ * pointer_size_;
+    public let node_size_ : Nat = Nat64.toNat(node_size);
+    public let leaf_size : Nat64 = key_size_ + value_size_;
+    public let root_size : Nat64 = root_aridity_ * pointer_size_;
+    public let offset_base : Nat64 = root_size - node_size;
+    public let padding : Nat64 = 8 - pointer_size_;
+    public let empty_values : Bool = args.value_size == 0;
+
+    public var leaf_count : Nat64 = 0;
+    public var node_count : Nat64 = 0;
+
+    public var storePointer : (offset : Nat64, child : Nat64) -> () = func(_, _) {};
+
+    public type State = {
+      nodes : Region;
+      leaves : Region;
+    };
+
+    var regions_ : ?State = null;
+
+    public func regions() : State {
+      switch (regions_) {
+        case (?r) r;
+        case (null) {
+          let nodes_region = Region.new();
+          let nodes : Region = {
+            region = nodes_region;
+            var freeSpace = 0;
+          };
+          let pages = (root_size + padding + 65536 - 1) / 65536;
+          assert Region.grow(nodes.region, pages) != 0xffff_ffff_ffff_ffff;
+          nodes.freeSpace := pages * 65536 - root_size - padding;
+          node_count := 1;
+
+          let leaves : Region = {
+            region = Region.new();
+            var freeSpace = 0;
+          };
+
+          let ret = { nodes = nodes; leaves = leaves };
+          regions_ := ?ret;
+          storePointer := switch (pointer_size_) {
+            case (8) func(offset, child) = Region.storeNat64(nodes_region, offset, child);
+            case (6) func(offset, child) {
+              Region.storeNat32(nodes_region, offset, Nat32.fromNat64(child & 0xffff_ffff));
+              Region.storeNat16(nodes_region, offset +% 4, Nat16.fromNat32(Nat32.fromNat64(child >> 32)));
+            };
+            case (5) func(offset, child) {
+              Region.storeNat32(nodes_region, offset, Nat32.fromNat64(child & 0xffff_ffff));
+              Region.storeNat8(nodes_region, offset +% 4, Nat8.fromNat16(Nat16.fromNat32(Nat32.fromNat64(child >> 32))));
+            };
+            case (4) func(offset, child) = Region.storeNat32(nodes_region, offset, Nat32.fromNat64(child));
+            case (2) func(offset, child) = Region.storeNat16(nodes_region, offset, Nat16.fromNat32(Nat32.fromNat64(child)));
+            case (_) Debug.trap("Can never happen");
+          };
+
+          ret;
+        };
+      };
+    };
+
+    // allocate can only be used for n <= 65536
```

Comment:
Looks like we need an assertion about leaf_size. Currently value_size is unlimited and can push leaf_size above this limit.